collector/fibrechannel_linux.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2017-2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build linux
+// +build !nofibrechannel
+
+package collector
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs/sysfs"
+)
+
+const maxUint64 = ^uint64(0)
+
+type fibrechannelCollector struct {
+	fs          sysfs.FS
+	metricDescs map[string]*prometheus.Desc
+	logger      log.Logger
+	subsystem   string
+}
+
+func init() {
+	registerCollector("fibrechannel", defaultEnabled, NewFibreChannelCollector)
+}
+
+// NewFibreChannelCollector returns a new Collector exposing FibreChannel stats.
+func NewFibreChannelCollector(logger log.Logger) (Collector, error) {
+	var i fibrechannelCollector
+	var err error
+
+	i.fs, err = sysfs.NewFS(*sysPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sysfs: %w", err)
+	}
+	i.logger = logger
+
+	// Detailed description for all metrics.
+	descriptions := map[string]string{
+		"dumped_frames_total":            "Number of dumped frames",
+		"loss_of_signal_total":           "Number of times signal has been lost",
+		"loss_of_sync_total":             "Number of failures on either bit or transmission word boundaries",
+		"rx_frames_total":                "Number of frames received",
+		"error_frames_total":             "Number of errors in frames",
+		"invalid_tx_word_total":          "Number of invalid words transmitted by host port",
```

Comment:
Probably should name this invalid_tx_word**s**_total

Reply:
Done!